### PR TITLE
[Doc] Add Python docstring for SNode

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2579

In addition to the docstring itself, this PR serves as an example for how to do class/method referencing.

### CODE REVIEW

File: python/taichi/lang/snode.py [hunk 1]
```diff
@@ -12,52 +12,119 @@
...
-        return SNode(self.ptr.bit_array(indices, dimensions, num_bits))
+            dimensions = [dimensions] * len(axes)
+        return SNode(self.ptr.bit_array(axes, dimensions, num_bits))
 
     def place(self, *args, offset=None, shared_exponent=False):
+        """Places a list of Taichi fields under the `self` container.
+
+        Args:
+            *args (List[ti.field]): A list of Taichi fields to place.
+            offsest (tupe[numbers.Number]): Offset of the field domain.
```
Comment (rexwangcc) [inline python/taichi/lang/snode.py:122]: It seems `offset` as an argument is `numbers.Number` than a `Tuple`?

File: python/taichi/lang/snode.py [hunk 2]
```diff
@@ -12,52 +12,119 @@
...
+        """
         if isinstance(dimensions, int):
-            dimensions = [dimensions] * len(indices)
-        return SNode(self.ptr.pointer(indices, dimensions))
+            dimensions = [dimensions] * len(axes)
+        return SNode(self.ptr.pointer(axes, dimensions))
 
-    def hash(self, indices, dimensions):
+    def hash(self, axes, dimensions):
+        """Not supported."""
```
Comment (rexwangcc) [inline python/taichi/lang/snode.py:48]: maybe we could add a deprecation warning later if this is not supported anymore?

**Review by rexwangcc — COMMENTED** [review]
I left a few comments but in general looks good!

One minor question: is the reason we don't use type annotations in `python/taichi/lang/snode.py` that it requires extra `import List, ... from typing`, or that it simply hasn't happened yet?

**Comment by k-ye** [comment]
> One minor question: is the reason we don't use type annotations in python/taichi/lang/snode.py that it requires extra import List, ... from typing, or that it simply hasn't happened yet?

It's a combination of me being lazy and the code hierarchy: I was planning to use the `Axis` type, but then ran into circular import issue...

**Comment by rexwangcc** [comment]
> > One minor question: is the reason we don't use type annotations in python/taichi/lang/snode.py that it requires extra import List, ... from typing, or that it simply hasn't happened yet?
> 
> It's a combination of me being lazy and the code hierarchy: I was planning to use the `Axis` type, but then ran into circular import issue...

Ah, gotcha. I'm fine with leaving them as is then! (we could use forward references to bypass the circular import, but I guess we can solve it once and for all by refactoring the code later?)

File: python/taichi/lang/snode.py (hunk 2 above)
Comment (k-ye) [reply python/taichi/lang/snode.py:48]: It's not exactly deprecated... Just stuck in a not supported mode when taichi switched to the LLVM backend

File: python/taichi/lang/snode.py (hunk 1 above)
Comment (k-ye) [reply python/taichi/lang/snode.py]: Actually it's a `Union[Number, tuple[Number]]` :)

File: python/taichi/lang/snode.py (hunk 2 above)
Comment (rexwangcc) [reply python/taichi/lang/snode.py:48]: I see!

**Review by rexwangcc — APPROVED** [review]
LGTM

**Comment by k-ye** [comment]
> (we could use forward references to bypass the circular import, but I guess we can solve it once and for all by refactoring the code later?)

+1, I feel like the code needs some refactor before we have all the types we need for annotation

**Pull request merged**